xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -178,6 +178,7 @@ public void onError(Status error) {
         //  a temporary solution. More design discussion needs to be done.
         if (error.getCode().equals(Code.NOT_FOUND)) {
           listener.onResult(ResolutionResult.newBuilder().build());
+          return;
         }
         listener.onError(error);
```

Comment:
I though it was intended in order to let client be able to know the exact error. I think a better way might be to populate the error to the empty resolution result like

```java
if (error.getCode().equals(Code.NOT_FOUND)) {
    listener.onResult(ResolutionResult.newBuilder()
        .setAttributes(Attributes.newBuilder().set(RESOLUTION_ERROR, error).build())
        .build());	
    return;       
}
```

@ejona86 any thought?

Reply:
This has something to do with the debate of whether we should fail RPCs immediately or continue using it when TD removes the resource that is currently being used. The discussion turns to be ambiguous and results in such a workaround in implementation. Most voices agree on (CDS/EDS) LB policies continue using Cluster/ClusterLoadAssignment information even if TD deletes it. Therefore, in CDS/EDS balancer's `handleNameResolutionError()` implementations, it's no-op is the policy currently has data to run. I don't have strong opinion on this. On the other hand, quite amount of people are arguing that RPCs should fail immediately if TD removes LDS/RDS resources. Since channel delegates name resolution error handling to LB policies, it would not work if the xDS resolver returns an error to the channel. Last time, we discussed this as our work around, to return an empty resolution result such that xDS LB policies will be offloaded and pick_first will turn channel into transient failure. I hope there could be a consistent decision for LB policies and resolver whether we should fail RPCs immediately if TD removes corresponding resources. In that case, we do not need this workaround to return an empty resolution result that forces channel go into transient failure.

Calling both `onResult()` and `onError()` doesn't seem to be a good solution, although the `onError()` after `onResult()` with empty resolution result does not have any behavior-wise operations other than showing the xDS resource not found message.

Reply:
I don't think `Attributes.newBuilder().set(RESOLUTION_ERROR, error).build()` would do anything. We don't want to each every policy about that.

Given where we stand today, it seems onResult (to change the policy) + onError (to report a precise error) seems fine. Although the status code should probably be converted to UNAVAILABLE. Just onResult by itself is okay, because it will typically become "NameResolver returned no usable address" which isn't the worst. I would like to improve how we handle things like this, but I don't think we should get bogged down trying to make this perfect at this point in time. We have other discussions in play to change how error handling is done.